lib/scanner/walk.go

```diff
@@ -98,30 +111,61 @@ type walker struct {
 
 // Walk returns the list of files found in the local folder by scanning the
 // file system. Files are blockwise hashed.
-func (w *walker) walk(ctx context.Context) (chan protocol.FileInfo, error) {
+func (w *walker) walk(ctx context.Context) (chan ScanResult, error) {
 	l.Debugln("Walk", w.Subs, w.BlockSize, w.Matcher)
 
 	if err := w.checkDir(); err != nil {
 		return nil, err
 	}
 
-	toHashChan := make(chan protocol.FileInfo)
-	finishedChan := make(chan protocol.FileInfo)
+	fsChan := make(chan fsWalkResult)
+
+	haveChan := make(chan *protocol.FileInfo)
+	haveCtx, haveCancel := context.WithCancel(ctx)
+
+	// A routine which walks the db and returns file infos to be compared
+	// to scan results.
+	go func() {
+		if len(w.Subs) == 0 {
+			w.Have.Walk("", haveCtx, haveChan)
+		} else {
+			haveCtxChan := haveCtx.Done()
```

Comment:
I don't think this loop is tight enough that it's worth temp-variabling the context channel, but it's a matter of taste.